src/coreclr/jit/importer.cpp

```diff
@@ -12967,28 +12970,31 @@ void Compiler::impInlineRecordArgInfo(InlineInfo*   pInlineInfo,
 #endif // FEATURE_SIMD
         }
 
-        // Spilling code relies on correct aliasability annotations.
-        assert(varDsc->lvHasLdAddrOp || varDsc->IsAddressExposed());
+        // Spilling code relies on correct aliasability annotations, except for
+        // retbufs that are not actually exposed in IL.
+        assert(varDsc->lvHasLdAddrOp || varDsc->IsAddressExposed() ||
+               (arg->GetWellKnownArg() == WellKnownArg::RetBuffer));
     }
 
-    if (curArgVal->gtFlags & GTF_ALL_EFFECT)
+    if (arg->GetWellKnownArg() == WellKnownArg::RetBuffer)
     {
-        inlCurArgInfo->argHasGlobRef = (curArgVal->gtFlags & GTF_GLOB_REF) != 0;
-        inlCurArgInfo->argHasSideEff = (curArgVal->gtFlags & (GTF_ALL_EFFECT & ~GTF_GLOB_REF)) != 0;
+        // Retbuffer is very often used directly in a dereference, and this
+        // generally makes it profitable to duplicate even when it is complex.
+        argInfo->argDuplicateComplex = true;
```

Comment:
This just minimizes some regressions, since the old behavior was more aggressive in duplicating the return buffer shape.